File: sdks/rust/src/bitmap.rs
//! Bitmap operations (SETBIT/GETBIT/BITCOUNT/BITPOS/BITOP)

use crate::client::SynapClient;
use crate::error::Result;
use serde::{Deserialize, Serialize};
use serde_json::json;

/// Bitmap operation types
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum BitmapOperation {
    And,
    Or,
    Xor,
    Not,
}

impl BitmapOperation {
    fn as_str(&self) -> &'static str {
        match self {
            BitmapOperation::And => "AND",
            BitmapOperation::Or => "OR",
            BitmapOperation::Xor => "XOR",
            BitmapOperation::Not => "NOT",
        }
    }
}

/// Bitmap statistics
#[derive(Debug, Clone, Serialize, Deserialize, Default)]
pub struct BitmapStats {
    pub total_bitmaps: usize,
    pub total_bits: usize,
    pub setbit_count: usize,
    pub getbit_count: usize,
    pub bitcount_count: usize,
    pub bitop_count: usize,
    pub bitpos_count: usize,
    pub bitfield_count: usize,
}

#[derive(Clone)]
pub struct BitmapManager {
    client: SynapClient,
}

impl BitmapManager {
    pub(crate) fn new(client: SynapClient) -> Self {
        Self { client }
    }

    /// Set bit at offset to value (SETBIT)
    ///
    /// # Arguments
    ///
    /// * `key` - Bitmap key
    /// * `offset` - Bit offset (0-based)
    /// * `value` - Bit value (0 or 1)
    ///
    /// # Returns
    ///
    /// Previous bit value (0 or 1)
    pub async fn setbit(&self, key: &str, offset: usize, value: u8) -> Result<u8> {
        if value > 1 {
            return Err(crate::error::SynapError::ServerError(
                "Bitmap value must be 0 or 1".to_string(),
            ));
        }

        let payload = json!({
            "key": key,
            "offset": offset,
            "value": value,
        });

        let response = self.client.send_command("bitmap.setbit", payload).await?;
        Ok(response["old_value"].as_u64().unwrap_or(0) as u8)
    }

    /// Get bit at offset (GETBIT)
    ///
    /// # Arguments
    ///
    /// * `key` - Bitmap key
    /// * `offset` - Bit offset (0-based)
    ///
    /// # Returns
    ///
    /// Bit value (0 or 1)
    pub async fn getbit(&self, key: &str, offset: usize) -> Result<u8> {
        let payload = json!({
            "key": key,
            "offset": offset,
        });

        let response = self.client.send_command("bitmap.getbit", payload).await?;
        Ok(response["value"].as_u64().unwrap_or(0) as u8)
    }

    /// Count set bits in bitmap (BITCOUNT)
    ///
    /// # Arguments
    ///
    /// * `key` - Bitmap key
    /// * `start` - Optional start offset (inclusive)
    /// * `end` - Optional end offset (inclusive)
    ///
    /// # Returns
    ///
    /// Number of set bits
    pub async fn bitcount(
        &self,
        key: &str,
        start: Option<usize>,
        end: Option<usize>,
    ) -> Result<usize> {
        let mut payload = json!({"key": key});
        if let Some(start_val) = start {
            payload["start"] = json!(start_val);
        }
        if let Some(end_val) = end {
            payload["end"] = json!(end_val);
        }

        let response = self.client.send_command("bitmap.bitcount", payload).await?;
        Ok(response["count"].as_u64().unwrap_or(0) as usize)
    }

    /// Find first bit set to value (BITPOS)
    ///
    /// # Arguments
    ///
    /// * `key` - Bitmap key
    /// * `value` - Bit value to search for (0 or 1)
    /// * `start` - Optional start offset (inclusive)
    /// * `end` - Optional end offset (inclusive)
    ///
    /// # Returns
    ///
    /// Position of first matching bit, or None if not found
    pub async fn bitpos(
        &self,
        key: &str,
        value: u8,
        start: Option<usize>,
        end: Option<usize>,
    ) -> Result<Option<usize>> {
        if value > 1 {
            return Err(crate::error::SynapError::ServerError(
                "Bitmap value must be 0 or 1".to_string(),
            ));
        }

        let mut payload = json!({
            "key": key,
            "value": value,
        });
        if let Some(start_val) = start {
            payload["start"] = json!(start_val);
        }
        if let Some(end_val) = end {
            payload["end"] = json!(end_val);
        }

        let response = self.client.send_command("bitmap.bitpos", payload).await?;

        if let Some(pos) = response["position"].as_u64() {
            Ok(Some(pos as usize))
        } else {
            Ok(None)
        }
    }

    /// Perform bitwise operation on multiple bitmaps (BITOP)
    ///
    /// # Arguments
    ///
    /// * `operation` - Bitwise operation (AND, OR, XOR, NOT)
    /// * `destination` - Destination key for result
    /// * `source_keys` - Source bitmap keys (NOT requires exactly 1 source)
    ///
    /// # Returns
    ///
    /// Length of resulting bitmap in bits
    ///
    /// # Errors
    ///
    /// Returns an error if:
    /// - NOT operation is used with more than one source key
    /// - No source keys provided
    pub async fn bitop<S>(
        &self,
        operation: BitmapOperation,
        destination: &str,
        source_keys: &[S],
    ) -> Result<usize>
    where
        S: AsRef<str>,
    {
        if operation == BitmapOperation::Not && source_keys.len() != 1 {
            return Err(crate::error::SynapError::ServerError(
                "NOT operation requires exactly one source key".to_string(),
            ));
        }

        if source_keys.is_empty() {
            return Err(crate::error::SynapError::ServerError(
                "BITOP requires at least one source key".to_string(),
            ));
        }

        let payload = json!({
            "destination": destination,
            "operation": operation.as_str(),
            "source_keys": source_keys.iter().map(|s| s.as_ref()).collect::<Vec<_>>(),
        });

        let response = self.client.send_command("bitmap.bitop", payload).await?;
        Ok(response["length"].as_u64().unwrap_or(0) as usize)
    }

    /// Execute bitfield operations (BITFIELD)
    ///
    /// # Arguments
    ///
    /// * `key` - Bitmap key
    /// * `operations` - Vector of bitfield operations
    ///
    /// # Returns
    ///
    /// Vector of result values (one per operation)
    ///
    /// # Example
    ///
    /// ```rust,no_run
    /// use synap_sdk::bitmap::BitmapManager;
    /// use serde_json::json;
    ///
    /// # async fn example(bitmap: BitmapManager) -> Result<(), Box<dyn std::error::Error>> {
    /// let operations = vec![
    ///     json!({
    ///         "operation": "SET",
    ///         "offset": 0,
    ///         "width": 8,
    ///         "signed": false,
    ///         "value": 42
    ///     }),
    ///     json!({
    ///         "operation": "GET",
    ///         "offset": 0,
    ///         "width": 8,
    ///         "signed": false
    ///     }),
    ///     json!({
    ///         "operation": "INCRBY",
    ///         "offset": 0,
    ///         "width": 8,
    ///         "signed": false,
    ///         "increment": 10,
    ///         "overflow": "WRAP"
    ///     }),
    /// ];
    ///
    /// let results = bitmap.bitfield("mybitmap", &operations).await?;
    /// # Ok(())
    /// # }
    /// ```
    pub async fn bitfield(&self, key: &str, operations: &[serde_json::Value]) -> Result<Vec<i64>> {
        let payload = json!({
            "key": key,
            "operations": operations,
        });

        let response = self.client.send_command("bitmap.bitfield", payload).await?;
        let results = response["results"]
            .as_array()
            .ok_or_else(|| {
                crate::error::SynapError::ServerError(
                    "Invalid response format for bitfield".to_string(),
                )
            })?
            .iter()
            .map(|v| v.as_i64().unwrap_or(0))
            .collect();

        Ok(results)
    }

    /// Retrieve bitmap statistics
    pub async fn stats(&self) -> Result<BitmapStats> {
        let response = self.client.send_command("bitmap.stats", json!({})).await?;
        let stats: BitmapStats = serde_json::from_value(response)?;
        Ok(stats)
    }
}

File: sdks/typescript/src/bitmap.ts
/**
 * Synap TypeScript SDK - Bitmap Manager
 * 
 * Redis-compatible Bitmap operations for bit-level manipulation
 */

import { SynapClient } from './client';
import type { CommandOptions } from './types';

/**
 * Bitmap operation types (AND, OR, XOR, NOT)
 */
export type BitmapOperation = 'AND' | 'OR' | 'XOR' | 'NOT';

/**
 * Bitmap statistics
 */
export interface BitmapStats {
  total_bitmaps: number;
  total_bits: number;
  setbit_count: number;
  getbit_count: number;
  bitcount_count: number;
  bitop_count: number;
  bitpos_count: number;
  bitfield_count: number;
}

interface SetBitResponse {
  key: string;
  old_value: number;
}

interface GetBitResponse {
  key: string;
  offset: number;
  value: number;
}

interface BitCountResponse {
  key: string;
  count: number;
}

interface BitPosResponse {
  key: string;
  position: number | null;
}

interface BitOpResponse {
  destination: string;
  length: number;
}

/**
 * Bitfield operation type
 */
export type BitfieldOperationType = 'GET' | 'SET' | 'INCRBY';

/**
 * Bitfield overflow behavior
 */
export type BitfieldOverflow = 'WRAP' | 'SAT' | 'FAIL';

/**
 * Bitfield operation specification
 */
export interface BitfieldOperation {
  operation: BitfieldOperationType;
  offset: number;
  width: number;
  signed?: boolean;
  value?: number; // Required for SET
  increment?: number; // Required for INCRBY
  overflow?: BitfieldOverflow; // For INCRBY, default: WRAP
}

interface BitfieldResponse {
  key: string;
  results: number[];
}

/**
 * Bitmap operations manager
 */
export class BitmapManager {
  constructor(private readonly client: SynapClient) {}

  private buildPayload(
    options: CommandOptions | undefined,
    extra: Record<string, unknown>
  ): Record<string, unknown> {
    const payload: Record<string, unknown> = { ...extra };

    if (options?.clientId) {
      payload.client_id = options.clientId;
    }

    return payload;
  }

  /**
   * Set bit at offset to value (SETBIT)
   * @param key Bitmap key
   * @param offset Bit offset (0-based)
   * @param value Bit value (0 or 1)
   * @param options Optional command options
   * @returns Previous bit value (0 or 1)
   */
  async setbit(
    key: string,
    offset: number,
    value: 0 | 1,
    options?: CommandOptions
  ): Promise<number> {
    if (value !== 0 && value !== 1) {
      throw new TypeError('Bitmap value must be 0 or 1');
    }

    const response = await this.client.sendCommand<SetBitResponse>(
      'bitmap.setbit',
      this.buildPayload(options, {
        key,
        offset,
        value,
      })
    );

    return response.old_value ?? 0;
  }

  /**
   * Get bit at offset (GETBIT)
   * @param key Bitmap key
   * @param offset Bit offset (0-based)
   * @param options Optional command options
   * @returns Bit value (0 or 1)
   */
  async getbit(key: string, offset: number, options?: CommandOptions): Promise<number> {
    const response = await this.client.sendCommand<GetBitResponse>(
      'bitmap.getbit',
      this.buildPayload(options, {
        key,
        offset,
      })
    );

    return response.value ?? 0;
  }

  /**
   * Count set bits in bitmap (BITCOUNT)
   * @param key Bitmap key
   * @param start Optional start offset (inclusive, default: 0)
   * @param end Optional end offset (inclusive, default: end of bitmap)
   * @param options Optional command options
   * @returns Number of set bits
   */
  async bitcount(
    key: string,
    start?: number,
    end?: number,
    options?: CommandOptions
  ): Promise<number> {
    const payload: Record<string, unknown> = { key };
    if (start !== undefined) {
      payload.start = start;
    }
    if (end !== undefined) {
      payload.end = end;
    }

    const response = await this.client.sendCommand<BitCountResponse>(
      'bitmap.bitcount',
      this.buildPayload(options, payload)
    );

    return response.count ?? 0;
  }

  /**
   * Find first bit set to value (BITPOS)
   * @param key Bitmap key
   * @param value Bit value to search for (0 or 1)
   * @param start Optional start offset (inclusive, default: 0)
   * @param end Optional end offset (inclusive, default: end of bitmap)
   * @param options Optional command options
   * @returns Position of first matching bit, or null if not found
   */
  async bitpos(
    key: string,
    value: 0 | 1,
    start?: number,
    end?: number,
    options?: CommandOptions
  ): Promise<number | null> {
    if (value !== 0 && value !== 1) {
      throw new TypeError('Bitmap value must be 0 or 1');
    }

    const payload: Record<string, unknown> = { key, value };
    if (start !== undefined) {
      payload.start = start;
    }
    if (end !== undefined) {
      payload.end = end;
    }

    const response = await this.client.sendCommand<BitPosResponse>(
      'bitmap.bitpos',
      this.buildPayload(options, payload)
    );

    return response.position ?? null;
  }

  /**
   * Perform bitwise operation on multiple bitmaps (BITOP)
   * @param operation Bitwise operation (AND, OR, XOR, NOT)
   * @param destination Destination key for result
   * @param sourceKeys Source bitmap keys (NOT requires exactly 1 source)
   * @param options Optional command options
   * @returns Length of resulting bitmap in bits
   */
  async bitop(
    operation: BitmapOperation,
    destination: string,
    sourceKeys: string[],
    options?: CommandOptions
  ): Promise<number> {
    if (operation === 'NOT' && sourceKeys.length !== 1) {
      throw new Error('NOT operation requires exactly one source key');
    }

    if (operation !== 'NOT' && sourceKeys.length < 1) {
      throw new Error('BITOP requires at least one source key');
    }

    const response = await this.client.sendCommand<BitOpResponse>(
      'bitmap.bitop',
      this.buildPayload(options, {
        destination,
        operation,
        source_keys: sourceKeys,
      })
    );

    return response.length ?? 0;
  }

  /**
   * Execute bitfield operations (BITFIELD)
   * @param key Bitmap key
   * @param operations List of bitfield operations
   * @param options Optional command options
   * @returns List of result values (one per operation)
   */
  async bitfield(
    key: string,
    operations: BitfieldOperation[],
    options?: CommandOptions
  ): Promise<number[]> {
    const response = await this.client.sendCommand<BitfieldResponse>(
      'bitmap.bitfield',
      this.buildPayload(options, {
        key,
        operations,
      })
    );

    return response.results ?? [];
  }

  /**
   * Retrieve bitmap statistics
   * @param options Optional command options
   * @returns Bitmap statistics
   */
  async stats(options?: CommandOptions): Promise<BitmapStats> {
    const response = await this.client.sendCommand<BitmapStats>(
      'bitmap.stats',
      this.buildPayload(options, {})
    );

    return response;
  }
}

